### ISSUE TITLE
loss calculation gap with Fig.12

### ISSUE THREAD

**Issue by MstarLioning** [SpatialVLA/SpatialVLA#49]
Hello! I’m sorry to trouble you again. We encountered a strange phenomenon when computing the loss using cross-entropy. Below is our calculation process, and we hope to receive your guidance.

## Calculation
```python
# 1) Pre-process inputs and run the model
processed_inputs = model_wrapper.processor(
    images=[resized_image_pil],
    text=[instruction],
    unnorm_key=model_wrapper.unnorm_key,
    suffix_actions=expert_action_7d_continuous.reshape(1, -1),  # 7-D continuous action
    return_tensors="pt",
    do_normalize=True,
)
model_outputs = pytorch_model(**processed_inputs)
logits = model_outputs.logits        # shape: [1, seq_len, 265347]

# 2) Ground-truth token IDs generated by the processor
labels = processed_inputs["labels"]  # shape: [1, 268]

# 3) Cross-Entropy loss (next-token prediction)
shift_logits = logits[..., :-1, :]   # [1, 267, 265347]
shift_labels = labels[..., 1:]       # [1, 267]

valid_mask   = shift_labels != -100
flat_logits  = shift_logits.view(-1, 265347)
flat_labels  = shift_labels.view(-1)

valid_logits = flat_logits[valid_mask.view(-1)]
valid_labels = flat_labels[valid_mask.view(-1)]

loss = torch.nn.functional.cross_entropy(valid_logits, valid_labels)
# loss_i = −log softmax(logits_i)[target_i];  final loss = mean(loss_i)

```

### 1.  RT1 Data Preprocessing 
We  implemented the RT1 dataset transformation:
```python
def apply_rt1_dataset_transform(raw_action_dict_np):
    gripper_action = raw_action_dict_np['gripper_closedness_action'][0]
    gripper_action_abs = rel2abs_gripper_actions(np.array([gripper_action]))[0]
    
    expert_action_7d_continuous = np.array([
        float(raw_action_dict_np['world_vector'][0]),     # x
        float(raw_action_dict_np['world_vector'][1]),     # y  
        float(raw_action_dict_np['world_vector'][2]),     # z
        float(raw_action_dict_np['rotation_delta'][0]),   # roll
        float(raw_action_dict_np['rotation_delta'][1]),   # pitch
        float(raw_action_dict_np['rotation_delta'][2]),   # yaw
        float(gripper_action_abs)                         # gripper (0→1 conversion)
    ], dtype=np.float32)
    
    return expert_action_7d_continuous
```

### 2.  Shift Logits Implementation  
We verified that our manual shift logits calculation matches the official implementation exactly:
```python
def compute_shift_logits_loss(model_outputs, labels):
    logits = model_outputs.logits
    shift_logits = logits[..., :-1, :]  # Remove last position
    shift_labels = labels[..., 1:]      # Remove first position
    
    # Results: Loss difference = 0.000000 
```

### 3.  **CORE ISSUE: Massive Quantization Errors**

Our token decoding reveals extremely large discretization errors:

**Example Step 1:**
```
Original 7D Action: [ 0.037274, 0.016109, -0.009015, 0.122997, -0.050189, -0.034994, 1.0]

Target Token IDs (from processor):
├─ Token 1 (Translation): Target ID = 259089
├─ Token 2 (Rotation):    Target ID = 263672  
├─ Token 3 (Sequence End): Target ID = 265346
└─ Token 4 (EOS):         Target ID = 1

Model Predictions :
├─ Token 1 (Translation): Predicted ID = 258033 (99.17% confidence) 
├─ Token 2 (Rotation):    Predicted ID = 262681 (99.80% confidence) 
├─ Token 3 (Sequence End): Predicted ID = 265346 (99.99% confidence) 
└─ Token 4 (EOS):         Predicted ID = [varies] (low confidence)  

Prediction Probabilities:
├─ P(Target Translation Token 259089) = 0.00000000 → Loss ≈ 15-20
├─ P(Target Rotation Token 263672)    = 0.00000056 → Loss ≈ 14.40
├─ P(Target Sequence End Token 265346) = 0.99999821 → Loss ≈ 0.0002
└─ P(Target EOS Token 1)              = 0.00000000 → Loss ≈ 15-20

Translation Token Misprediction:
├─ Target: 259089 → Our decoded action: (x=0.285, y=0.152, z=0.032)
├─ Predicted: 258033 → Our decoded action: (x=0.278, y=0.000, z=0.155)  
├─ Token ID difference: |259089 - 258033| = 1,056 tokens apart
└─ Continuous action error: Δx=0.248, Δy=0.136, Δz=0.041

Rotation Token Misprediction:
├─ Target: 263672 → Our decoded action: (roll=0.188, pitch=-0.063, yaw=0.063)
├─ Predicted: 262681 → Our decoded action: (roll=0.000, pitch=-0.188, yaw=-0.125)
├─ Token ID difference: |263672 - 262681| = 991 tokens apart  
└─ Continuous action error: Δroll=0.065, Δpitch=0.012, Δyaw=0.097

```
**Resulting Loss: 15.8988**

## Environment
- Model: `IPEC-COMMUNITY/spatialvla-4b-224-pt`
- Dataset: `fractal20220817_data/0.1.0`
- Implementation: Using `SpatialVLAInference` with `google_robot` policy setup

## Our Understanding of SpatialVLA's Action Tokenization

Based on our code analysis and the paper, we understand that SpatialVLA uses:

1. **3 tokens per action** (not 7): Translation + Rotation + Gripper
2. **Polar coordinate system** for translation: (θ, φ, r) → 16×32×8 = 4,096 tokens
3. **Euler angles** for rotation: (roll, pitch, yaw) → 16×16×16 = 4,096 tokens  
4. **Binary gripper**: open/close → 2 tokens
5. **RT1 dataset transform**: Applies `rel2abs_gripper_actions()` preprocessing

**Comment by DelinQu** [comment]
Hi,
I'm so sorry for my late reply. Here are the loss/accuracy metrics when fine-tuning SpatialVLA on the Fractal dataset. The model maintains high accuracy and low L1 error.
<img width="1181" alt="Image" src="https://github.com/user-attachments/assets/491386b2-138f-4e8b-a84e-c467a011fc51" />

However, potential problems might lie in: `apply_rt1_dataset_transform`, `Normalization` and `data augmentation`. In [processing_spatialvla.py#L216](https://github.com/SpatialVLA/SpatialVLA/blob/f8904f5962e9bf0fe733e7abadd65b02ae1658b4/model/processing_spatialvla.py#L219C9-L219C19), we use unnorm_key to denormalize model outputs to match GT action scale. However, when computing token IDs for GT actions at [processing_spatialvla.py#L108](https://github.com/SpatialVLA/SpatialVLA/blob/f8904f5962e9bf0fe733e7abadd65b02ae1658b4/model/processing_spatialvla.py#L108), the code assumes actions are already normalized. Could you clarify your GT action normalization process?

**Comment by MstarLioning** [comment]
@DelinQu 您好！非常感谢您的回复和建议！我们根据您提到的几个问题分别check和修正了一下，目前已经实现Loss与Accuracy与您提供图片一样的数量级了，但是虽然数量级接近，可是具体的数值在10000个样本上的计算仍高于您展示的图表（L1在0.0475、交叉熵在1.27、准确率在73%、Grad_Norm的范围波动极大，存在于0.002~138.00之间）。故而想请教您一下您使用的模型是IPEC-COMMUNITY/spatialvla-4b-224-pt还是IPEC-COMMUNITY/spatialvla-4b-224-sft-fractal呢？（我们当前采用的是COMMUNITY/spatialvla-4b-224-pt，切换为IPEC-COMMUNITY/spatialvla-4b-224-sft-fractal后似乎损失进一步提升）或者不知道我们有没有什么遗漏的细节？期待您的进一步指导！

![Image](https://github.com/user-attachments/assets/183c6ce7-c1dc-49a6-9710-ea7139cb7c42)
图1：计算得到的Loss以及Accuracy

![Image](https://github.com/user-attachments/assets/52b35c96-bc20-49db-b5ee-509a3a581403)
图2：单一Step表现，未预测正确但正确答案处于Top5的位置

![Image](https://github.com/user-attachments/assets/f1c67652-e4c5-416d-84fa-cdbfc90f4f7a)
图3：计算得到的梯度，基本都在两位数左右的水平，与您图里grad_norm所展示的1.几相距很远

![Image](https://github.com/user-attachments/assets/021cc292-f451-4aab-9f73-cf4c3be291bf)
图4：IPEC-COMMUNITY/spatialvla-4b-224-sft-fractal在10000个样本下的结果

**Comment by DelinQu** [comment]
这个loss是从COMMUNITY/spatialvla-4b-224-pt微调到IPEC-COMMUNITY/spatialvla-4b-224-sft-fractal，初期loss略有波动几十个epoch后会稳定在高ACC。我这边check一下上传的权重，确实是sft-fractal微调的结果。我还是倾向于说数据预处理，图像增强等等可能没完全对齐。

**Comment by MstarLioning** [comment]
收到，我会仔细检查并尝试向您的代码对齐。再次感谢您的指导！

**Comment by OrangeSodahub** [comment]
@DelinQu 请问后续情况怎么样了？谢谢！

**Comment by MstarLioning** [comment]
@OrangeSodahub 您也遇见相同的问题了么？我自己目前还没有解决
